// @angular-extensions/elements/bundles/angular-extensions-elements-testing.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@angular-extensions/elements/testing', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['angular-extensions'] = global['angular-extensions'] || {}, global['angular-extensions'].elements = global['angular-extensions'].elements || {}, global['angular-extensions'].elements.testing = {}), global.ng.core));
}(this, (function (exports, core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * Generated from: lazy-element-testing.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var LazyElementTestingDirective = /** @class */ (function () {
        /**
         * @param {?} vcr
         * @param {?} template
         */
        function LazyElementTestingDirective(vcr, template) {
            this.vcr = vcr;
            this.template = template;
        }
        /**
         * @return {?}
         */
        LazyElementTestingDirective.prototype.ngOnInit = function () {
            this.vcr.createEmbeddedView(this.template);
        };
        return LazyElementTestingDirective;
    }());
    LazyElementTestingDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[axLazyElement]',
                },] }
    ];
    /** @nocollapse */
    LazyElementTestingDirective.ctorParameters = function () { return [
        { type: core.ViewContainerRef },
        { type: core.TemplateRef }
    ]; };
    if (false) {
        /**
         * @type {?}
         * @private
         */
        LazyElementTestingDirective.prototype.vcr;
        /**
         * @type {?}
         * @private
         */
        LazyElementTestingDirective.prototype.template;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lazy-element-dynamic-testing.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var LazyElementDynamicTestingDirective = /** @class */ (function () {
        /**
         * @param {?} vcr
         * @param {?} template
         */
        function LazyElementDynamicTestingDirective(vcr, template) {
            this.vcr = vcr;
            this.template = template;
        }
        /**
         * @return {?}
         */
        LazyElementDynamicTestingDirective.prototype.ngOnInit = function () {
            this.vcr.createEmbeddedView(this.template);
        };
        return LazyElementDynamicTestingDirective;
    }());
    LazyElementDynamicTestingDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[axLazyElementDynamic]',
                },] }
    ];
    /** @nocollapse */
    LazyElementDynamicTestingDirective.ctorParameters = function () { return [
        { type: core.ViewContainerRef },
        { type: core.TemplateRef }
    ]; };
    if (false) {
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicTestingDirective.prototype.vcr;
        /**
         * @type {?}
         * @private
         */
        LazyElementDynamicTestingDirective.prototype.template;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lazy-elements-testing.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var LazyElementsTestingModule = /** @class */ (function () {
        function LazyElementsTestingModule() {
        }
        return LazyElementsTestingModule;
    }());
    LazyElementsTestingModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [
                        LazyElementTestingDirective,
                        LazyElementDynamicTestingDirective,
                    ],
                    exports: [LazyElementTestingDirective, LazyElementDynamicTestingDirective],
                },] }
    ];

    /**
     * @fileoverview added by tsickle
     * Generated from: public_api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: angular-extensions-elements-testing.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.LazyElementDynamicTestingDirective = LazyElementDynamicTestingDirective;
    exports.LazyElementTestingDirective = LazyElementTestingDirective;
    exports.LazyElementsTestingModule = LazyElementsTestingModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
