// @angular-extensions/elements/esm2015/lib/lazy-elements/lazy-elements.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/lazy-elements/lazy-elements.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, NgModule, Optional, SkipSelf, } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LazyElementDirective } from './lazy-element/lazy-element.directive';
import { LazyElementDynamicDirective } from './lazy-element-dynamic/lazy-element-dynamic.directive';
import { LazyElementsLoaderService, } from './lazy-elements-loader.service';
import { LAZY_ELEMENT_CONFIGS, LAZY_ELEMENT_ROOT_GUARD, LAZY_ELEMENT_ROOT_OPTIONS, } from './lazy-elements.tokens';
/**
 * @param {?} rootOptions
 * @return {?}
 */
export function createLazyElementRootGuard(rootOptions) {
    if (rootOptions) {
        throw new TypeError(`LazyElementsModule.forRoot() called twice. Feature modules should use LazyElementsModule.forFeature() instead.`);
    }
    return 'LazyElementsModule.forRoot() multiple execution guard';
}
export class LazyElementsModule {
    /**
     * @param {?} lazyElementsLoaderService
     * @param {?} elementConfigsMultiProvider
     * @param {?} guard
     */
    constructor(lazyElementsLoaderService, elementConfigsMultiProvider, guard) {
        if (elementConfigsMultiProvider && elementConfigsMultiProvider.length) {
            elementConfigsMultiProvider
                .filter((/**
             * @param {?} configs
             * @return {?}
             */
            (configs) => configs.some((/**
             * @param {?} config
             * @return {?}
             */
            (config) => !config.isAdded))))
                .forEach((/**
             * @param {?} configs
             * @return {?}
             */
            (configs) => lazyElementsLoaderService.addConfigs(configs)));
        }
    }
    /**
     * @param {?} options
     * @return {?}
     */
    static forRoot(options) {
        return {
            ngModule: LazyElementsModule,
            providers: [
                {
                    provide: LAZY_ELEMENT_CONFIGS,
                    useValue: options && options.elementConfigs ? options.elementConfigs : [],
                    multi: true,
                },
                {
                    provide: LAZY_ELEMENT_ROOT_OPTIONS,
                    useValue: options.rootOptions ? options.rootOptions : {},
                },
                {
                    provide: LAZY_ELEMENT_ROOT_GUARD,
                    useFactory: createLazyElementRootGuard,
                    deps: [[LAZY_ELEMENT_ROOT_OPTIONS, new Optional(), new SkipSelf()]],
                },
            ],
        };
    }
    /**
     * @param {?} options
     * @return {?}
     */
    static forFeature(options) {
        return {
            ngModule: LazyElementsModule,
            providers: [
                {
                    provide: LAZY_ELEMENT_CONFIGS,
                    useValue: options && options.elementConfigs ? options.elementConfigs : [],
                    multi: true,
                },
            ],
        };
    }
}
LazyElementsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [LazyElementDirective, LazyElementDynamicDirective],
                imports: [CommonModule],
                exports: [LazyElementDirective, LazyElementDynamicDirective],
                providers: [],
            },] }
];
/** @nocollapse */
LazyElementsModule.ctorParameters = () => [
    { type: LazyElementsLoaderService },
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [LAZY_ELEMENT_CONFIGS,] }] },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [LAZY_ELEMENT_ROOT_GUARD,] }] }
];
/**
 * @record
 */
export function LazyElementModuleOptions() { }
if (false) {
    /** @type {?|undefined} */
    LazyElementModuleOptions.prototype.elementConfigs;
}
/**
 * @record
 */
export function LazyElementModuleRootOptions() { }
if (false) {
    /** @type {?|undefined} */
    LazyElementModuleRootOptions.prototype.elementConfigs;
    /** @type {?|undefined} */
    LazyElementModuleRootOptions.prototype.rootOptions;
}
/**
 * @record
 */
export function LazyElementRootOptions() { }
if (false) {
    /** @type {?|undefined} */
    LazyElementRootOptions.prototype.loadingComponent;
    /** @type {?|undefined} */
    LazyElementRootOptions.prototype.errorComponent;
    /** @type {?|undefined} */
    LazyElementRootOptions.prototype.isModule;
    /** @type {?|undefined} */
    LazyElementRootOptions.prototype.importMap;
    /** @type {?|undefined} */
    LazyElementRootOptions.prototype.preload;
    /** @type {?|undefined} */
    LazyElementRootOptions.prototype.hooks;
}
