// @angular-extensions/elements/esm2015/lib/lazy-elements/lazy-elements.tokens.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/lazy-elements/lazy-elements.tokens.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export const LAZY_ELEMENT_CONFIGS = new InjectionToken('LAZY_ELEMENT_CONFIGS');
/** @type {?} */
export const LAZY_ELEMENT_ROOT_OPTIONS = new InjectionToken('LAZY_ELEMENT_ROOT_OPTIONS');
/** @type {?} */
export const LAZY_ELEMENT_ROOT_GUARD = new InjectionToken('LAZY_ELEMENT_ROOT_GUARD');
/** @type {?} */
export const LAZY_ELEMENTS_REGISTRY = new InjectionToken('LAZY_ELEMENTS_REGISTRY', {
    providedIn: 'root',
    factory: (/**
     * @return {?}
     */
    () => new Map()),
});
/**
 * @record
 */
export function LazyElementsRegistry() { }
if (false) {
    /** @type {?} */
    LazyElementsRegistry.prototype.get;
    /** @type {?} */
    LazyElementsRegistry.prototype.set;
    /** @type {?} */
    LazyElementsRegistry.prototype.has;
}
