// @angular-extensions/elements/fesm2015/angular-extensions-elements-testing.js
import { Directive, ViewContainerRef, TemplateRef, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * Generated from: lazy-element-testing.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LazyElementTestingDirective {
    /**
     * @param {?} vcr
     * @param {?} template
     */
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementTestingDirective.decorators = [
    { type: Directive, args: [{
                selector: '[axLazyElement]',
            },] }
];
/** @nocollapse */
LazyElementTestingDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    LazyElementTestingDirective.prototype.vcr;
    /**
     * @type {?}
     * @private
     */
    LazyElementTestingDirective.prototype.template;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lazy-element-dynamic-testing.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LazyElementDynamicTestingDirective {
    /**
     * @param {?} vcr
     * @param {?} template
     */
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementDynamicTestingDirective.decorators = [
    { type: Directive, args: [{
                selector: '[axLazyElementDynamic]',
            },] }
];
/** @nocollapse */
LazyElementDynamicTestingDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    LazyElementDynamicTestingDirective.prototype.vcr;
    /**
     * @type {?}
     * @private
     */
    LazyElementDynamicTestingDirective.prototype.template;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lazy-elements-testing.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LazyElementsTestingModule {
}
LazyElementsTestingModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    LazyElementTestingDirective,
                    LazyElementDynamicTestingDirective,
                ],
                exports: [LazyElementTestingDirective, LazyElementDynamicTestingDirective],
            },] }
];

/**
 * @fileoverview added by tsickle
 * Generated from: public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: angular-extensions-elements-testing.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { LazyElementDynamicTestingDirective, LazyElementTestingDirective, LazyElementsTestingModule };
